### PR TITLE
fix: tweak aggregated resource allocation history endpoint

### PR DESCRIPTION
**Pull request by dzhu**
## Description

- Add missing query SQL file, without which it didn't work at all.
- Format dates as YYYY-MM-DD in gRPC responses.
- Support monthly aggregation.

## Test Plan

- [x] carefully examine the list of unstaged files in my working tree
- [x] run the query on some test data and check that the dates come out right

### CODE REVIEW

File: master/internal/db/postgres_proto.go [hunk 1]
```diff
@@ -38,18 +39,32 @@ func protoParser(rows *sqlx.Rows, val interface{}) error {
...
 		switch parsed := value.(type) {
 		case []byte:
 			var marshaled interface{}
-			if err := json.Unmarshal(parsed, &marshaled); err != nil {
+			if err = json.Unmarshal(parsed, &marshaled); err != nil {
 				return errors.Wrapf(err, "error parsing field: %s", key)
 			}
 			dest[key] = marshaled
+		case time.Time:
+			if colTypes[key] == "DATE" {
```
Comment (stoksc) [inline master/internal/db/postgres_proto.go:65]: if the column type for this key isn't `DATE`, it seems like we should error or something, not just continue without scanning a value.

File: master/internal/db/postgres_proto.go [hunk 2]
```diff
@@ -38,18 +39,32 @@ func protoParser(rows *sqlx.Rows, val interface{}) error {
...
 	for key, value := range dest {
 		switch parsed := value.(type) {
 		case []byte:
 			var marshaled interface{}
-			if err := json.Unmarshal(parsed, &marshaled); err != nil {
+			if err = json.Unmarshal(parsed, &marshaled); err != nil {
 				return errors.Wrapf(err, "error parsing field: %s", key)
 			}
 			dest[key] = marshaled
+		case time.Time:
```
Comment (stoksc) [inline master/internal/db/postgres_proto.go:64]: could you just to_char the date in SQL to get around needing this? I think this is what has been done historically to make protoParser happy.  I don't really like that.. at all. Just curious what you think about protoParser in general.

File: master/internal/core.go [hunk 3]
```diff
@@ -218,45 +218,62 @@ func (m *Master) getRawResourceAllocation(c echo.Context) error {
...
 	resp := &apiv1.ResourceAllocationAggregatedResponse{}
-	if err := m.db.QueryProto(
-		"get_aggregated_allocation", &resp.ResourceEntries, start.UTC(), end.UTC(),
-	); err != nil {
-		return errors.Wrap(err, "error fetching allocation data")
-	}
 
-	c.Response().Header().Set("Content-Type", "text/csv")
+	switch args.Period {
+	case "RESOURCE_ALLOCATION_AGGREGATION_PERIOD_DAILY":
```
Comment (stoksc) [inline master/internal/core.go:232]: Could you convert args.Period to a `masterv1.ResourceAllocationAggregationPeriod` and then de-dup this code with the code in `api_master.go`?

**Review by stoksc — COMMENTED** [review]
looks good, just some q's

File: master/internal/core.go (hunk 3 above)
Comment (dzhu) [reply master/internal/core.go]: Great idea, done!

File: master/static/srv/get_monthly_aggregated_allocation.sql [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        ELSE 'RESOURCE_ALLOCATION_AGGREGATION_TYPE_UNSPECIFIED'
+    END AS aggregation_type,
+    resource_aggregates.aggregation_key,
+    sum(resource_aggregates.seconds) AS seconds
+FROM
+    resource_aggregates,
+    const
+WHERE
+    -- `@>` determines whether the range contains the time.
+    const.period @> resource_aggregates.date
```
Comment (stoksc) [inline master/static/srv/get_monthly_aggregated_allocation.sql:21]: did we consider a btree index on resource_aggregates.date? I think i remember the conversation somewhere.. if there are N users, M labels, this is what, N + M rows per date? If you query for a month maybe ~10^3 rows, after a year maybe ~10^4 rows in the table. By dumb math, it makes me think an index could speed up by a factor of 10 after this exists a while?

**Review by stoksc — APPROVED** [review]
looks great.

File: master/static/srv/get_monthly_aggregated_allocation.sql (hunk 4 above)
Comment (dzhu) [reply master/static/srv/get_monthly_aggregated_allocation.sql:21]: Sounds good, I'll add that and land.

Comment (dzhu) [reply master/static/srv/get_monthly_aggregated_allocation.sql:21]: Oh, there's already a unique constraint, which [automatically adds an index](https://www.postgresql.org/docs/10/indexes-unique.html).

**Pull request merged**